3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,248 @@
+    this.sentinel_ = null;
+
+    /** @private {Object<string, AmpAnalytics3pCreativeMessageRouter>} */
+    this.creativeMessageRouters_ = [];
+
+    /**
+     * Simple requestIdleCallback polyfill
+     * @type {!function(!function(), number)}
+     */
+    this.requestIdleCallback =
+      this.win_.requestIdleCallback.bind(this.win_) ||
+      (cb => setTimeout(cb, 1));
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    try {
+      const frameNameData = JSON.parse(this.win_.name);
+      if (frameNameData.sentinel) {
+        this.sentinel_ = frameNameData.sentinel;
+        this.iframeMessagingClient_.setSentinel(this.sentinel_);
+      }
+    } catch (e) {
+      dev().warn(TAG_, 'Unable to set sentinel');
+      return;
+    }
+    this.iframeMessagingClient_.registerCallback(
+      MessageTypes.ampAnalytics3pMessages,
+      messages => {
+        // All the messages are for this frame, but may come from different
+        // creatives.
+        messages[MessageTypes.ampAnalytics3pMessages].forEach(msg => {
+          if (!this.creativeMessageRouters_[msg.senderId]) {
+            this.creativeMessageRouters_[msg.senderId] =
+              new AmpAnalytics3pCreativeMessageRouter(this, this.sentinel_,
+                msg.senderId);
+          }
+          this.creativeMessageRouters_[msg.senderId].receiveMessage(msg);
+        });
+        Object.entries(this.creativeMessageRouters_).forEach(
+          entry => {
+            const creativeMessageRouter = entry[1];
+            creativeMessageRouter.sendQueuedMessagesToListeners();
+          });
+      });
+    this.sendReadyMessageToCreative_();
+  }
+
+  /**
+   * Gets a handle to the window (or testing object that mocks window)
+   * @returns {!Window}
+   */
+  getWindow() {
+    return this.win_;
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @private
+   */
+  sendReadyMessageToCreative_() {
+    this.iframeMessagingClient_.sendMessage(MessageTypes.ampAnalytics3pReady,
+      {senderId: this.sentinel_});
+  }
+
+  /**
+   * Sends a message back to the host window
+   * @param {string} type The type of message to send.
+   * @param {Object=} opt_payload The payload of message to send.
+   */
+  sendMessage(type, opt_payload) {
+    this.iframeMessagingClient_.sendMessage(type, opt_payload);
+  }
+}
+
+/**
+ * @private @const {!AmpAnalytics3pMessageRouter}
+ */
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
+   * @param {!string} sentinel
+   * @param {!string} creativeId
+   */
+  constructor(parent, sentinel, creativeId) {
+    /** @private {!Window} */
```

Comment:
Yep. Was originally this.win_ but made a last-minute change to parent and didn't update the typing. Fixed.